src/main/java/com/es/phoneshop/model/product/Cart/HttpSessionCartService.java

```diff
@@ -0,0 +1,65 @@
+package com.es.phoneshop.model.product.Cart;
+
+import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.exceptions.OutOfStockException;
+import com.es.phoneshop.model.product.exceptions.ProductNotFoundException;
+
+import javax.servlet.http.HttpServletRequest;
+import java.math.BigDecimal;
+
+public class HttpSessionCartService implements CartService {
+    protected static final String CART_SESSION_ATTRIBUTE = HttpSessionCartService.class + ".cart";
+
+    private HttpSessionCartService() {
+    }
+
+    public static HttpSessionCartService getInstance() {
+        return SingletonHolder.instance;
+    }
+
+    public static class SingletonHolder {
+        public static final HttpSessionCartService instance = new HttpSessionCartService();
+    }
+
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
+        Cart result = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (result == null) {
+            result = new Cart();
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, result);
+        }
+        return result;
+    }
+
+    @Override
+    public void add(Cart cart, Product product, int quantity) throws OutOfStockException {
+        CartItem currentCartItem = cart.getCartItems().stream()
+                .filter(x -> x.getProductId().equals(product.getId()))
+                .findFirst()
+                .orElse(null);
```

Comment:
Don't use orElse(null) when using Java 8 optionals. Use the optional instead.
int currentProductQuantity = currentCartItemOptional
    .map(CartItem::getQuantity)
    .orElse(0);
And so on.